Package.swift

```diff
@@ -1,11 +1,19 @@
-// swift-tools-version:5.3
+// swift-tools-version:5.9
 import PackageDescription
 
 let clibsodiumTarget: Target
 #if os(OSX) || os(macOS) || os(tvOS) || os(watchOS) || os(iOS)
     clibsodiumTarget = .binaryTarget(
         name: "Clibsodium",
         path: "Clibsodium.xcframework")
+#elseif os(Windows)
+    clibsodiumTarget = .target(
+        name: "Clibsodium",
+        path: "Clibsodium-win",
```

Comment:
Can we use the path `Clibsodium`? I think that the headers should be identical across the platforms, so I'm not sure that the `-win` suffix adds much. Additionally, the import library, DLL, DSO, and static library would be different enough that we should be able to differentiate between them. 
A subsequent step would be to convert this to a nuget package, which would only further stratify the layout. The nuget packaging is important as we are using that already for consuming other packages and is definitely something that we would want - it avoids checking in binaries into the repository.

Finally, I think that we should change the compilation here. We should use `/Zi` when compiling to allow us to merge and unique the information without having to worry about clang vs MSVC.

Reply:
Thank you for those comments, super insightful!
I do have a few followup questions, I'm sorry if there are stupid, but then they are even more important as it means my mental model is off and I'm just building it ;) :
 1. How do we feel about making this tailored made for our usecase vs. upstreamable? - I imagine this format where we just drop `.lib` into the repo would be acceptable by upstream as it's very similar to how they handle macos. I'm not sure how would we "sell" the nuget packet solution to upstream spm repo?
 2. When you say we should use `/Zi` do you mean we should use our own compilation of `libsodium` rather then "offical" release libs? I'm not opinionated about that - just asking. I think we may not be able to work around that anyway - as the official release doesn't ship `arm64` which IIUC we want to support.
 3. When you mention DLL, DSO and static, is the goal here that the package for windows should support all 3? Is this something we actively use and need to have or a "nice to have" for the future?

I totally agree with `Clibsodium-win` having separate include is bad, I'm just not familiar with `xcframework` so I didn't want to risk any extra problems that early in development of this package. Totally something to be fixed.

Reply:
> Thank you for those comments, super insightful! I do have a few followup questions, I'm sorry if there are stupid, but then they are even more important as it means my mental model is off and I'm just building it ;) :
> 
> 1. How do we feel about making this tailored made for our usecase vs. upstreamable? - I imagine this format where we just drop `.lib` into the repo would be acceptable by upstream as it's very similar to how they handle macos. I'm not sure how would we "sell" the nuget packet solution to upstream spm repo?

I think that setting up the image for consumption in a very "standard" (there are $n$ standards, lets make one to unify them, great now there are $n + 1$ standards 😮‍💨) manner is better because it simplifies the build process for others.  We already have a few external dependencies. They are all setup as:

~~~
[name]-[version]
  +- usr
     +- bin
     +- lib
     +- include
~~~

As a concrete example, `firebase-development\usr\lib\firebase.lib`. This pattern is replicated so it is easier for others to know where to look to find the components.

As to the upstream SPM repo, that is easy to sell: nuget _is_ one of the sanctioned system package formats. It is also cross-platform so we can use it on other platforms if desired.

The other point to consider: we are building with CMake and we already have infrastructure for fetching and versioning with nuget releases. If that means that we need to re-package for our consumption, that doesn't seem too terrible. We have firebase, sentry, WinAppSDK, and ArcCore already consumed as nuget packages.

> 2. When you say we should use `/Zi` do you mean we should use our own compilation of `libsodium` rather then "offical" release libs? I'm not opinionated about that - just asking. I think we may not be able to work around that anyway - as the official release doesn't ship `arm64` which IIUC we want to support.

Yes, we should use our own build of the product as the PDB specification is not well known (it is mostly reverse engineered). As such, clang and LLVM only partially understand the debug information, and I'm concerned what impact this might have if we start feeding it a mismatched PDB to merge. The embedded codeview representation (`/Z7`) will ensure that we do not choke on anything we don't know as the linker will reference the specific bits from the object file rather than try to merge the PDB. This also avoids a PDBSRV dependency.

> 3. When you mention DLL, DSO and static, is the goal here that the package for windows should support all 3? Is this something we actively use and need to have or a "nice to have" for the future?

Having the ability to select between the format would be a nice to have. A DLL and DSO are the same representation, just different terminology for platforms (DLL = Windows, DSO = Linux, DYLIB = macOS). Having static linking for the module should be fine is we are able to contain it. I tend to be wary of third party library linkage because you can easily run afoul of ODR violations.

> I totally agree with `Clibsodium-win` having separate include is bad, I'm just not familiar with `xcframework` so I didn't want to risk any extra problems that early in development of this package. Totally something to be fixed.

I don't think that we should touch the xcframework on macOS. XCFrameworks are limited to Xcode, and non-Xcode platforms and builds are not permitted to use that as per upstream. Use of artifactbundle is not possible as that is not universally supported. As such, I think that using nuget packages for non-Xcode builds remain the most reasonable approach.

Reply:
Thank you!

One more question about upstream. What dependency do we expect upstream to accept? Dependency on our instance of nuget? Or do we expect them to accept the new release pipeline just to support nuget?

In any case, what I gathered from that as well as discussion with Brian is my next steps here are:

 1) Build a `libsodium.lib` for `arm64` windows and check if there are any problems with that.
 2) Make the cmake definition for the same package using "commited in" `lib` files.
 3) This should open merging and using `swift-sodium` within `arc`
 4) Prepare build of `libsodium` that does:
      1) Build with `/Zi` for both `arm64` and `x64`
      2) Prepares nuget package and release it
 5) Optionally: Add support for dynamic linking.
 6) Attempt to upstream that.

Reply:
I think that we should determine if they are open to releasing nuget packages. If so, we just upstream the release workflow changes required to package and release the nupkg. I think that Sentry was not willing to do that until they could test on ARM64, so we have our own rules for packaging.

I think that the `/Z7` issue is important, as that impacts debug information. If you want to hold off on that, we should drop the PDB and say that this code is not possible to debug, build your own and wire that up (which many of the product engineers might find daunting). However, once we are building it - its pretty trivial to package it up into a nuget package. If the issue is of experience, we can pair on the task - it is pretty trivial and shouldn't take more than an afternoon (depending on the build times for libsodium).

I think that the dynamic linking and upstreaming both being after the merging is reasonable.

Reply:
Probably worth changing the name of the path to `libsodium`.